lib/wallet/src/Cardano/Wallet/Flavor.hs

```diff
@@ -67,13 +75,41 @@ instance WalletFlavor (RndAnyState n p) where
 instance WalletFlavor (SharedState n SharedKey) where
     walletFlavor = SharedWallet
 
+-- | A type for states that will be used in tests.
 newtype TestState s (k :: (Depth -> Type -> Type)) = TestState s
     deriving (Generic, Show, Eq)
 
+-- | A type family to get the key type from a state.
 type family KeyOf (s :: Type) :: (Depth -> Type -> Type) where
     KeyOf (SeqState n k) = k
     KeyOf (RndState n) = ByronKey
     KeyOf (SharedState n k) = k
     KeyOf (SeqAnyState n k p) = k
     KeyOf (RndAnyState n p) = ByronKey
     KeyOf (TestState s k) = k
+
+-- | A singleton type to capture the flavor of a key.
+data KeyFlavorS a where
+    ByronKeyS :: KeyFlavorS ByronKey
+    IcarusKeyS :: KeyFlavorS IcarusKey
+    ShelleyKeyS :: KeyFlavorS ShelleyKey
+    SharedKeyS :: KeyFlavorS SharedKey
+
+-- | Map a wallet flavor to a key flavor.
+keyOfWallet :: WalletFlavorS s -> KeyFlavorS (KeyOf s)
+keyOfWallet ShelleyWallet = ShelleyKeyS
+keyOfWallet IcarusWallet = IcarusKeyS
+keyOfWallet ByronWallet = ByronKeyS
+keyOfWallet SharedWallet = SharedKeyS
+keyOfWallet BenchByronWallet = ByronKeyS
+keyOfWallet BenchShelleyWallet = ShelleyKeyS
+
+-- | A function to reify the flavor of a key from a state type.
+--
+-- use with
+-- > keyFlavor @s
+keyFlavor :: forall s. WalletFlavor s => KeyFlavorS (KeyOf s)
+keyFlavor = keyOfWallet (walletFlavor @s)
+
+-- | Constraints for a state with a specific key.
+type StateWithKey s k = (WalletFlavor s, KeyOf s ~ k)
```

Comment:
```suggestion
```

Reply:
Will remove on next PR, not to stress CI, thanks for spotting it